### DISCUSSION TITLE
LFX Mentorship 2023 01-Mar-May Challenge - for #2229

### DISCUSSION THREAD

**Discussion by juntao** [WasmEdge/WasmEdge#2230] [Show and tell]
## Introduction

For those who want to apply for LFX mentorship for #2229, this is a pretest for the application. The goal of this LFX mentorship is to create a Rust library crate for mediapipe models to run on WasmEdge NN. We designed this challenge to evaluate the candidates.

## Requirements

1 [Install WasmEdge with Tensorflow Lite plugin and dependency libraries](https://wasmedge.org/book/en/write_wasm/rust/wasinn.html#get-wasmedge-with-wasi-nn-plug-in-tensorflow-lite-backend)

2 [Try the mobilenet TFLite model example in Rust](https://wasmedge.org/book/en/write_wasm/rust/wasinn.html#tensorflow-lite-backend-example)

3 Write a similar demo for the [Mediapipe Face Detection](https://google.github.io/mediapipe/solutions/face_detection.html) TFLite model. The output should be a list of detected face boundary coordinates, and landmark coordinates on the detected faces.

4 Once you are finished, please provide your link to the GitHub repository, which must contain a README file to show how you do and some screenshots of the examples' execution results.

## Deadline

Please refer to [the timeline of LFX mentorship 2023 01-Mar-May](https://github.com/cncf/mentoring/tree/main/lfx-mentorship/2023/01-Mar-May).
Since we have to process the internal review and decide, the final date for valid submissions will be February 20, 5:00 PM PDT.

**Comment by Hrushi20** [comment]
@juntao, Can we start working on the Pretest?

**Comment by sarrah-basta** [comment]
Hello @juntao , I would like to apply for LFX mentorship for https://github.com/WasmEdge/WasmEdge/issues/2229 as I mentioned in the introduction at https://github.com/WasmEdge/WasmEdge/issues/2229#issuecomment-1399278784 and have hence started working on this pretest. 

I have already **successfully completed the first two requirements** that outline the setup and usecase of the given task, along with trying some simple Rust applications that run in WasmEdge and am starting on the main task in requirement 3. Any guidance regarding other steps I need to do would be highly appreciated !

**Reply by juntao** [reply]
That's great. Thank you. Look forward to seeing your work!

**Comment by Hrushi20** [comment]
Hello, 
I've used the Face detection model to get the required anchors. The outputs for the short-range [model](https://google.github.io/mediapipe/solutions/models.html#face-detection) are 896 * 16. I need to perform a Non-Maximum Suppression algorithm on these anchor boxes. Do I need to write the code to generate the [anchors](https://d2l.ai/chapter_computer-vision/anchor.html) along with [Non-Maximum Suppression](https://d2l.ai/chapter_computer-vision/anchor.html#predicting-bounding-boxes-with-non-maximum-suppression) algorithm from scratch or is there a package in WasmEdge which does the post processing of the data.

**Reply by juntao** [reply]
Yes, for the internship task, you will need to create a Rust impl of NMS. I believe it is not that difficult. 

https://towardsdatascience.com/non-maximum-suppression-nms-93ce178e177c

Alternatively, you could create WasmEdge host functions for the OpenCV library, which has most of these utilities implemented in C already. 

But for the purpose of this pre-test, you do not have to do that. Just pick a box at random and draw it. Thanks.

**Comment by sarrah-basta** [comment]
Hello @juntao , 
I have used the face detection model and gotten the output as well. However, I am trying to ascertain whether the obtained output is correct and reached to some concerns. 

1. The specified [model](https://drive.google.com/file/d/1d4-xJP9PVzOvMBDgIjz6NhvpnlG9_i0S/preview) takes in a normalized image as input, and since I could not find any functions to normalize an image in the [image](https://docs.rs/image/0.24.5/image/index.html) crate, should this function be written from scratch or instead is it okay to use another image processing crate such as [opencv::core](https://docs.rs/opencv/latest/opencv/core/fn.normalize.html) ?
2. Similarly, the output of the model is expected in two parts : 
- [{'name': 'regressors', 'index': 175, 'shape': array([  1, 896,  16], dtype=int32) }: an array of 896 * 16 * 4 bytes  with information about the detected faces and
- {'name': 'classificators', 'index': 174, 'shape': array([  1, 896,   1], dtype=int32)}: an array of 896 * 1 * 4 bytes with raw detection scores 

These are then used for the NMS algorithm, and I understand we can pick a box at random. However, since the get_output function only returns a buffer with 57344 bytes = 896 * 16 of data, I think it only gets the first output, and wanted to know if I should only use the [wasi-nn crate](https://docs.rs/wasi-nn/0.2.1/src/wasi_nn/generated.rs.html#120) as it does not exactly have support for tensorflow-lite yet, or should something like [wasmedge-tensorflow-interface](https://docs.rs/wasmedge_tensorflow_interface/0.2.2/wasmedge_tensorflow_interface/fn.wasmedge_tensorflowlite_get_output_tensor.html) be explored instead ?

**Reply by juntao** [reply]
You can use the `image_to_tensor` example code below to normalize the input image. 

https://wasmedge.org/book/en/write_wasm/rust/wasinn.html#using-wasi-nn-with-tensorflow-lite-backend-in-rust

For the output tensor, it is your job to figure out what it means. You could refer to the Python code in mediapipe to figure it out. Thanks.

**Comment by Hrushi20** [comment]
@juntao, The [graph](https://github.com/mengfu188/simple-mediapipe/blob/master/mediapipe/graphs/face_detection/face_detection_mobile_cpu.pbtxt) contains the required steps for post-processing. So, can we compile [mediapipe tflite module](https://github.com/google/mediapipe/tree/master/mediapipe/calculators/tflite) written in cpp to webassembly and call it during execution. This would help us not rewrite the code. The module contains all the required functionalities for post-processing. Currently I'm stuck with creating the anchors to find detection of face.

**Reply by juntao** [reply]
That's a great idea. It is possible to compile c/c++ code together with Rust to create a Wasm file. An example is how to created the wasmedge_quickjs project. We wrapped a Rust API around a C library.

**Comment by sarrah-basta** [comment]
Hello @juntao ,

I have completed the requirements of the challenge to the best of my understanding and uploaded the **executable** `.wasm` files along with a detailed README describing the outputs obtained at this github repository : https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo.git .

My entire code to obtain the _list of detected face boundary coordinates, and landmark coordinates on the detected faces_ is written in `rust`, and I have added a `python` script that can be used to visualize these outputs as a temporary solution. More details are described in the [README.md](https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo/blob/master/README.md) file. 

Looking forward to your feedback!

**Comment by yanghaku** [comment]
Hello! This is my [repository](https://github.com/yanghaku/mediapipe-wasinn-demo) for the pretest.  Looking forward to your feedback, thank you.

**Comment by sarrah-basta** [comment]
> Hello @juntao ,
> 
> I have completed the requirements of the challenge to the best of my understanding and uploaded the **executable** `.wasm` files along with a detailed README describing the outputs obtained at this github repository : https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo.git .
> 
> My entire code to obtain the _list of detected face boundary coordinates, and landmark coordinates on the detected faces_ is written in `rust`, and I have added a `python` script that can be used to visualize these outputs as a temporary solution. More details are described in the [README.md](https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo/blob/master/README.md) file.
> 
> Looking forward to your feedback!

Greetings ! As the date for the submission of the pretest was extended, I updated my code by adding a *rust implementation of NMS* and *modularized my code*.  The updated code can be found along with a detailed [README.md](https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo/blob/master/README.md) file describing the [outputs obtained](https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo#similar-results-can-be-obtained-by-replacing-pathtoimage_filejpg-in-both-the-commands-with-the-other-two-images-in-the-images-folder:~:text=To%20visualize%20these%20facial%20keypoints%20on%20the%20image%2C%20the%20output%20image%20can%20be%20found%20at%20images/test_image.jpg_drawn_out.jpg%20path.) at the same github repository as earlier : https://github.com/sarrah-basta/wasmedge-wasinn-mediapipe-facedet-demo.git .

**Reply by juntao** [reply]
Hi @sarrah-basta Does that mean your solution will no longer depend on Python? Thanks.

**Reply by sarrah-basta** [reply]
Yes @juntao, while I was earlier using python to draw the keypoints on the image, I was able to use the [imageproc ](https://docs.rs/imageproc/0.18.0/imageproc/drawing/fn.draw_hollow_rect.html)crate for that too! So now there is no dependency on python.
